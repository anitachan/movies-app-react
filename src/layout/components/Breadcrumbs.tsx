import { Person } from '@mui/icons-material';
import ArtTrackIcon from '@mui/icons-material/ArtTrack';
import FavoriteIcon from '@mui/icons-material/Favorite';
import HomeIcon from '@mui/icons-material/Home';
import NavigateNextIcon from '@mui/icons-material/NavigateNext';
import { Link, Breadcrumbs as MuiBreadcrumbs, SvgIconTypeMap, Typography } from '@mui/material';
import { OverridableComponent } from '@mui/material/OverridableComponent';
import { useLocation } from 'react-router-dom';

const breadcrumbNameMap: Record<
  string,
  { text: string; icon: OverridableComponent<SvgIconTypeMap> }
> = {
  '/home': { text: 'Home', icon: HomeIcon },
  '/favorites': { text: 'Favorites', icon: FavoriteIcon },
  '/details': { text: 'Details', icon: ArtTrackIcon },
  '/actor': { text: 'Actor', icon: Person },
};

export const Breadcrumbs = () => {
  const location = useLocation();
  const pathNames = location.pathname.split('/').filter((x) => x);

  return (
    <MuiBreadcrumbs
      separator={<NavigateNextIcon fontSize="small" />}
      aria-label="breadcrumb"
      sx={{
        color: 'white',
        pl: 2,
        pt: 0.5,
      }}
    >
      <Typography variant="h6" noWrap component="div">
        <Link underline="none" color="inherit">
          Movies
        </Link>
      </Typography>
      {pathNames.map((value, index) => {
        const to = `/${pathNames.slice(0, index + 1).join('/')}`;

        const prevSegment = pathNames[index - 1];
        const isMovieIdSegment = prevSegment === 'details' || prevSegment === 'actor';

        if (isMovieIdSegment) return null;

        const matchKey = Object.keys(breadcrumbNameMap).find((key) => to.startsWith(key));
        const { icon: Icon, text } = breadcrumbNameMap[matchKey ?? ''] ?? {
          icon: HomeIcon,
          text: value,
        };

        return (
          <Typography
            key={to}
            color="white"
            sx={{ display: 'flex', alignItems: 'center', gap: 0.5 }}
          >
            <Icon fontSize="small" />
            {text}
          </Typography>
        );
      })}
    </MuiBreadcrumbs>
  );
};
