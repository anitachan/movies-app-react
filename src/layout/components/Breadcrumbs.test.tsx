import { render, screen } from '@testing-library/react';
import { Breadcrumbs } from './Breadcrumbs';

const mockLocation = {
  pathname: '/foo/bar',
  search: '',
  hash: '',
  state: null,
  key: 'test',
};

jest.mock(
  'react-router-dom',
  () => ({
    useLocation: () => mockLocation,
  }),
  { virtual: true }
);

describe('Breadcrumbs', () => {
  test('should create and contain the current route', () => {
    const { container } = render(<Breadcrumbs />);
    const route = screen.getByText('foo');

    expect(container).toMatchSnapshot();
    expect(route).toBeInTheDocument();
  });

  test('skips rendering the ID segment when previous segment is "details"', () => {
    mockLocation.pathname = '/details/123';

    render(<Breadcrumbs />);

    expect(screen.getByText('Details')).toBeInTheDocument();
    expect(screen.queryByText('123')).toBeNull();
  });
});
