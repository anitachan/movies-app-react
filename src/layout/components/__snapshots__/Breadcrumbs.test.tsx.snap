// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should create and contain the current route 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1d5226r-MuiTypography-root-MuiBreadcrumbs-root"
  >
    <ol
      class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
    >
      <li
        class="MuiBreadcrumbs-li"
      >
        <div
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-5lnhk4-MuiTypography-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-6b04s2-MuiTypography-root-MuiLink-root"
          >
            Movies
          </a>
        </div>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
          data-testid="NavigateNextIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-17ymgc2-MuiTypography-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
            data-testid="HomeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          foo
        </p>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
          data-testid="NavigateNextIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-17ymgc2-MuiTypography-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
            data-testid="HomeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          bar
        </p>
      </li>
    </ol>
  </nav>
</div>
`;
